File: packages/volt-git/src/reporter.ts
/**
 * CLI progress reporter — turns streamed `ProgressFrame`s into stderr output. On a TTY it overwrites a single
 * live line; when captured/piped (the AI tool, CI logs) it emits throttled discrete lines so there's a trace
 * without spam. stdout is never touched, so `--json` output stays clean. `finish()` clears the TTY line before
 * the command prints its own summary.
 */
import type { ProgressHandler } from "./bridge/types.js";

export interface Reporter {
	onProgress: ProgressHandler;
	finish: () => void;
}

/** Prefix for the machine-readable progress line a GUI host (the VS Code extension) opts into via
 *  `VOLT_PROGRESS_JSON=1`. It parses `VOLT_PROGRESS <json>` off stderr to drive a real progress bar. */
const PROGRESS_JSON_PREFIX = "VOLT_PROGRESS ";

export function createReporter(): Reporter {
	// A GUI host sets VOLT_PROGRESS_JSON=1 to get structured frames on stderr (parseable); humans/AI get the
	// pretty/throttled text form. stdout is never touched either way.
	if (process.env.VOLT_PROGRESS_JSON === "1") {
		return {
			onProgress: (p) => process.stderr.write(`${PROGRESS_JSON_PREFIX}${JSON.stringify(p)}\n`),
			finish: () => {},
		};
	}

	const tty = process.stderr.isTTY === true;
	let lastBucket = -1;
	let lastLabel = "";
	let wroteTty = false;

	const onProgress: ProgressHandler = (p) => {
		const pct = p.total && p.total > 0 ? Math.floor((p.done / p.total) * 100) : undefined;
		const label = p.phase ?? p.operation;
		const line = pct !== undefined ? `${label}: ${p.done}/${p.total} (${pct}%)` : `${label}…`;
		if (tty) {
			process.stderr.write(`\r\x1b[2K${line}`); // carriage-return + clear-line → overwrite in place
			wroteTty = true;
		} else {
			// Non-TTY: emit only when the 10% bucket or the label changes — a readable trace, not a flood.
			const bucket = pct !== undefined ? Math.floor(pct / 10) : -1;
			if (bucket !== lastBucket || label !== lastLabel) {
				process.stderr.write(`${line}\n`);
				lastBucket = bucket;
				lastLabel = label;
			}
		}
	};

	const finish = (): void => {
		if (tty && wroteTty) process.stderr.write("\r\x1b[2K"); // clear the live line for the summary
	};

	return { onProgress, finish };
}

File: packages/volt-cli/src/Volt.Cli/Sync/Reporter.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Volt.Cli.Core.Wire;

namespace Volt.Cli.Sync;

/// <summary>CLI progress reporter — streamed ProgressFrames → stderr (stdout stays clean for --json). A GUI host
/// sets VOLT_PROGRESS_JSON=1 to get structured frames it parses; humans/AI get throttled text. C# port of
/// the original TypeScript implementation</summary>
public static class Reporter
{
    private const string ProgressJsonPrefix = "VOLT_PROGRESS ";
    private static readonly JsonSerializerOptions Json = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public static Action<ProgressFrame> Create()
    {
        if (Environment.GetEnvironmentVariable("VOLT_PROGRESS_JSON") == "1")
            return p => Console.Error.Write(ProgressJsonPrefix + JsonSerializer.Serialize(p, Json) + "\n");

        var lastBucket = -1;
        var lastLabel = "";
        return p =>
        {
            int? pct = p.Total is > 0 ? (int)Math.Floor((double)p.Done / p.Total.Value * 100) : null;
            var label = p.Phase ?? p.Operation;
            var line = pct is not null ? $"{label}: {p.Done}/{p.Total} ({pct}%)" : $"{label}…";
            var bucket = pct ?? -1;
            if (bucket / 10 != lastBucket / 10 || label != lastLabel)
            {
                Console.Error.WriteLine(line);
                lastBucket = bucket;
                lastLabel = label;
            }
        };
    }
}
